### PR TITLE
Expand docutils.parsters.rst.states

### PR DESCRIPTION
**Pull request by adamtheturtle**
I used stubgen and broke out a few classes.

### CODE REVIEW

File: stubs/docutils/docutils/parsers/rst/states.pyi [hunk 1]
```diff
@@ -6,11 +6,58 @@ from typing import Any
 from typing_extensions import TypeAlias
 
 from docutils import nodes
+from docutils.statemachine import StateMachine, StateWS
 from docutils.utils import Reporter
 
 class Struct:
     def __init__(self, **keywordargs) -> None: ...
 
+class RSTState(StateWS):
```
Comment (adamtheturtle) [inline stubs/docutils/docutils/parsers/rst/states.pyi:15]: This is taken from `stubgen` output except where specified.

File: stubs/docutils/docutils/parsers/rst/states.pyi [hunk 2]
```diff
@@ -6,11 +6,58 @@ from typing import Any
...
+    nested_sm: type[StateMachine]
```
Comment (adamtheturtle) [inline stubs/docutils/docutils/parsers/rst/states.pyi:16]: I added this type not given by `stubgen`.

File: stubs/docutils/docutils/parsers/rst/states.pyi [hunk 3]
```diff
@@ -6,11 +6,58 @@ from typing import Any
...
+    nested_sm_cache: Incomplete
+    nested_sm_kwargs: Incomplete
+    def __init__(self, state_machine, debug: bool = False) -> None: ...
+    memo: Incomplete
+    reporter: Incomplete
+    inliner: Incomplete
+    document: nodes.document
```
Comment (adamtheturtle) [inline stubs/docutils/docutils/parsers/rst/states.pyi:23]: I added this type not given by `stubgen`.

**Comment by adamtheturtle** [comment]
Looking at the primer output:

```diff
+ sphinx/ext/autodoc/directive.py: note: In member "run" of class "AutodocDirective":
+ sphinx/ext/autodoc/directive.py:118:30: error: "Reporter" has no attribute "get_source_and_line"  [attr-defined]
```

The code is https://github.com/sphinx-doc/sphinx/blob/f85f50e52614c627adbcf43629ca028829454081/sphinx/ext/autodoc/directive.py#L117, and it appears that the authors expect an `AttributeError` in some cases:

```python
        try:
            source, lineno = reporter.get_source_and_line(
                self.lineno)
        except AttributeError:
            source, lineno = (None, None)
```

I'm not sure what to do about:

```diff
+ sphinx/ext/inheritance_diagram.py: note: In member "run" of class "InheritanceDiagram":
+ sphinx/ext/inheritance_diagram.py:389:57: error: Argument 5 has incompatible type "RSTState"; expected "Inliner"  [arg-type]
```

Also, this will require users (Sphinx, in the output) to give generic types type parameters.

**Comment by adamtheturtle** [comment]
Pinging you here as thanks for taking a look at my first foray into `TypeVar`s 😉

**Review by srittau — COMMENTED** [review]
I have no idea how docutils works internally, but I'm a bit wary of all the newly generic classes, especially `Directive`. Unless you are sure this is correct, I would keep `Directive` non-generic for now, and just use `RSTState[Incomplete]` as annotate for now.

**Comment by adamtheturtle** [comment]
Thank you @srittau . I have made the change I believe you wanted.

**Review by srittau — APPROVED** [review]
Thanks!

**Pull request merged**